modules/platform/nuxeo-drive-server/nuxeo-drive-mongodb/src/main/java/org/nuxeo/drive/mongodb/MongoDBAuditChangeFinder.java

```diff
@@ -0,0 +1,179 @@
+import org.nuxeo.ecm.core.query.sql.model.Predicates;
+import org.nuxeo.ecm.core.query.sql.model.QueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.AuditQueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.AuditReader;
+import org.nuxeo.ecm.platform.audit.api.LogEntry;
+import org.nuxeo.mongodb.audit.MongoDBAuditBackend;
+import org.nuxeo.mongodb.audit.MongoDBAuditEntryReader;
+import org.nuxeo.runtime.api.Framework;
+
+import com.mongodb.client.MongoCollection;
+
+/**
+ * Override the JPA audit based change finder to execute query in BSON.
+ * <p>
+ * The structure of the query executed by the {@link AuditChangeFinder} is:
+ *
+ * <pre>
+ * from LogEntry log where log.repositoryId = :repositoryId
+ *
+ * + AND if ActiveRoots (activeRoots) NOT empty
+ *
+ * from LogEntry log where log.repositoryId = :repositoryId and (
+ * LIST_DOC_EVENTS_IDS_QUERY and ( ROOT_PATHS or COLECTIONS_PATHS) or
+ * (log.category = 'NuxeoDrive' and log.eventId != 'rootUnregistered') )
+ *
+ *
+ * if ActiveRoots EMPTY:
+ *
+ * from LogEntry log where log.repositoryId = :repositoryId and ((log.category =
+ * 'NuxeoDrive' and log.eventId != 'rootUnregistered'))
+ *
+ * + AND (log.id > :lowerBound and log.id <= :upperBound) + order by
+ * log.repositoryId asc, log.eventDate desc
+ * </pre>
+ *
+ * @since 11.3
+ */
+public class MongoDBAuditChangeFinder extends AuditChangeFinder {
+
+    private static final Logger log = LogManager.getLogger(MongoDBAuditChangeFinder.class);
+
+    @Override
+    public long getUpperBound() {
+        RepositoryManager repositoryManager = Framework.getService(RepositoryManager.class);
+        return getUpperBound(new HashSet<>(repositoryManager.getRepositoryNames()));
+    }
+
+    /**
+     * Returns the last available log id in the audit index considering events older than the last clustering
+     * invalidation date if clustering is enabled for at least one of the given repositories. This is to make sure the
+     * {@code DocumentModel} further fetched from the session using the audit entry doc id is fresh.
+     */
+    @Override
+    public long getUpperBound(Set<String> repositoryNames) {
+        long clusteringDelay = getClusteringDelay(repositoryNames);
+        MongoDBAuditBackend auditService = (MongoDBAuditBackend) Framework.getService(AuditReader.class);
+        QueryBuilder queryBuilder = new AuditQueryBuilder().predicate(
+                Predicates.in(LOG_REPOSITORY_ID, repositoryNames));
+        if (clusteringDelay > -1) {
+            // Double the delay in case of overlapping, see https://jira.nuxeo.com/browse/NXP-14826
+            long lastClusteringInvalidationDate = System.currentTimeMillis() - 2 * clusteringDelay;
+            queryBuilder.and(Predicates.lt(LOG_LOG_DATE, lastClusteringInvalidationDate));
+        }
+        queryBuilder.order(OrderByExprs.desc(LOG_ID));
+        queryBuilder.limit(1);
+        List<LogEntry> entries = auditService.queryLogs(queryBuilder);
```

Comment:
Have you checked the impact of such an ordered query on a huge collection?

I'm referring to the crucial [optimization](https://github.com/nuxeo/nuxeo/blob/master/modules/platform/nuxeo-drive-server/nuxeo-drive-elasticsearch/src/main/java/org/nuxeo/drive/elasticsearch/ESAuditChangeFinder.java#L251) that was done for the Elasticsearch change finder in the context of https://jira.nuxeo.com/browse/NXP-24232..

Reply:
No, I didn't think about that aspect. I need to check that thanks !

Reply:
It happens the reverted index is kinda built in. This explains why one can sort or desc _id easily.
Tested it with 10 millions documents anyway and the result was instant.